### PR TITLE
feat: sequencer auto recover when meet an unexpected shutdown

### PR DESCRIPTION
**Pull request by krish-nr**
### Description

This PR aims to fix the issue where the sequencer node fails to recover after a crash (specifically when the sequencer is a PBSS node).
related node [PR](https://github.com/bnb-chain/opbnb/pull/250)

### Rationale

When a sequencer node crashes, it may fail to persist the journal in time. As a result, when Geth is restarted, the journal data cannot be read, leading to the loss of recent state data. This causes the sequencer to fail during the buildPayload process, rendering it unable to continue operating. The diagram below illustrates the sequencer block production flow, with the red sections highlighting the logic that cannot proceed due to the crash.

In the diagram, sequencerAction alternates between stages (1) and (2). In stage (1), the payload is constructed, and in stage (2), the data is persisted. This process is not synchronously blocked; in stage (1), operations such as filling the payload's transactions (txs) and some other tasks are performed asynchronously, while the payload is returned synchronously and immediately, controlled by a condition lock (cond). Therefore, before the `update` in stage (1) is completed, the `getPayload` in stage (2) will be in a `cond.wait` state.

After the sequencer crashes, due to the loss of state data (for example, the sequencer crashes at block height 34123456 and the next block to be built is 34123457), the `prepareWork` phase depends on the state data of block 34123456. However, after the crash, the state data of block 34123456 is lost, leading to failure in this phase and consequently preventing the system from entering the `update` logic. As a result, the process will remain indefinitely blocked at `getPayload`, unable to make progress.

image

The fix involves adding two routines to handle the recovery process and monitor the recovery progress. Upon a failure in the generate phase, a fix routine is started based on specific error conditions. To avoid blocking the main process, a separate routine is also initiated to monitor the specific block being repaired. Once the data recovery is complete, a retry of the update process is triggered, allowing the system to recover the state from before the crash and continue making progress.

There are two scenarios for recovery: recovering from local data or from peers (for the sequencer, peers are its backup nodes). In most cases, the data can be recovered locally. However, there is a corner case where local recovery fails: if the sequencer has already gossiped the latest block to peer nodes but crashes during the local persistence process, the sequencer may fall behind the peer by one block. In this extreme situation, the sequencer must recover from the peer.

This is the complete sequencer recovery flow, as illustrated below.

image

### Example

add an example CLI or API response...

### Changes

Notable changes:
* fix_manager object has been added to the worker to manage the status and progress during the fix process.
* Added a method to recover the state from local data.
* Introduced a new downloader method in the Backend interface.

### CODE REVIEW

File: miner/fix_manager.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// FixManager manages the fix operation state and notification mechanism.
+type FixManager struct {
```
[the remark this answers is not in the document]
Comment (krish-nr) [reply miner/fix_manager.go]: updated

File: miner/payload_building.go [hunk 2]
```diff
@@ -348,8 +392,17 @@ func (w *worker) buildPayload(args *BuildPayloadArgs) (*Payload, error) {
 			start := time.Now()
 			// getSealingBlock is interrupted by shared interrupt
 			r := w.getSealingBlock(fullParams)
+
+			// if state missing, init fixing routine
+			if r.err != nil && strings.Contains(r.err.Error(), "missing trie node") {
+				log.Info("step into fixing")
+				w.StartFix(payload.id, fullParams.parentHash)
+				return 0
```
[the remark this answers is not in the document]
Comment (krish-nr) [reply miner/payload_building.go]: updated

File: core/blockchain.go [hunk 3]
```diff
@@ -242,16 +242,17 @@ type BlockChain struct {
...
+	snaps                 *snapshot.Tree                   // Snapshot tree for fast trie leaf access
+	triegc                *prque.Prque[int64, common.Hash] // Priority queue mapping block numbers to tries to gc
+	gcproc                time.Duration                    // Accumulates canonical block processing for trie dumping
+	lastWrite             uint64                           // Last block when the state was flushed
+	flushInterval         atomic.Int64                     // Time interval (processing time) after which to flush a state
+	triedb                *triedb.Database                 // The database handler for maintaining trie nodes.
+	stateCache            state.Database                   // State database to reuse between imports (contains state cache)
+	proofKeeper           *ProofKeeper                     // Store/Query op-proposal proof to ensure consistent.
+	txIndexer             *txIndexer                       // Transaction indexer, might be nil if not enabled
+	stateRecoveringStatus bool
```
[the remark this answers is not in the document]
Comment (krish-nr) [reply core/blockchain.go]: updated

File: miner/fix_manager.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+func NewFixManager() *StateFixManager {
+	return &StateFixManager{}
+}
+
+// StartFix launches a goroutine to manage the fix process and tracks the fix state.
+func (fm *StateFixManager) StartFix(worker *worker, id engine.PayloadID, parentHash common.Hash) error {
+	fm.mutex.Lock()
+	defer fm.mutex.Unlock()
+
+	if fm.isFixInProgress {
```
Comment (owen-reorg) [inline miner/fix_manager.go:28]: Seems there is no place setting isFixInProgress to true.

**Review by andyzhang2023 — APPROVED** [review]
LGTM

**Review by owen-reorg — APPROVED** [review]

**Pull request merged**